com.unity.ml-agents/Runtime/Sensors/RayPerceptionSensor.cs

```diff
@@ -314,12 +356,14 @@ public int Write(ObservationWriter writer)
             {
                 Array.Clear(m_Observations, 0, m_Observations.Length);
                 var numRays = m_RayPerceptionInput.Angles.Count;
-                var numDetectableTags = m_RayPerceptionInput.DetectableTags.Count;
+                var rayObservartionSize = GetObservationSizePerRay();
 
                 // For each ray, write the information to the observation buffer
                 for (var rayIndex = 0; rayIndex < numRays; rayIndex++)
                 {
-                    m_RayPerceptionOutput.RayOutputs?[rayIndex].ToFloatArray(numDetectableTags, rayIndex, m_Observations);
+                    Array.Clear(m_SingleRayObservations, 0, rayObservartionSize);
+                    RayOutputToArray(m_RayPerceptionOutput.RayOutputs[rayIndex], rayIndex, m_SingleRayObservations);
+                    Array.Copy(m_SingleRayObservations, 0, m_Observations, rayObservartionSize * rayIndex, rayObservartionSize);
```

Comment:
Changes here are debatable since it hurts performance. I introduced `m_SingleRayObservations` to avoid having to handle the buffer offset for each ray in `RayOutputToArray()` but it will create more array copy and clear operations. I couldn't find a way to create a reference to subarray to avoid copying.

One alternatives would be having `RayOutputToArray` to take all the `m_RayPerceptionOutput.RayOutputs` and the whole `m_Observations` array as input and let user handle the buffer offset themselves.

Reply:
I think instead of this, you can just call `writer.AddList` once per ray. That probably lets you get rid of m_observations entirely.

Reply:
* and pass an offset to `writer.AddList`

Reply:
yeah you're right, didn't notice that